arch: arc: necessary fixes for smp

vonhust opened the pull request:
After  recent changes in Zephyr 's SMP high-level code and tests cases,  the arc smp doesn't work and failed in the tests related to SMP and multi-core.

This PR fixes all the failures.

Fixes #19885 
Fixes #19599

Commit:
arch: arc: necessary fixes after normal idle is used for SMP

* necessary fixes after commit 11bd67db where ipi interrupt is used
to notify other cores to do a thread switch if necessary

* then for arc, it's needed to ignore swap_ok and check whether thread
switch is needed in the exit of irq handling.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: do not use sleep instruction for nsim smp

It's found that in nsim_hs_smp, sometimes the cpu
doesn't response inter-core interrupt after executing sleep
instruction.

It may be a bug of nsim, but needs more time to
investigate the root of this issue.

This commit is a workround for this, as nsim is just an
instruction simulator, no direct impact.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: split codes for SMP and codes for multicore

reported by #19599, this commit splits the codes for Zephyr SMP
and codes for ARC mulicore.

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

andrewboie commented on arch/arc/core/cpu_idle.S:
```diff
@@ -43,7 +43,18 @@ SECTION_FUNC(TEXT, z_arch_cpu_idle)
 	ld r1, [z_arc_cpu_sleep_mode]
 	or r1, r1, (1 << 4) /* set IRQ-enabled bit */
+/*
+ * It's found that (in nsim_hs_smp and hsdk), when cpu
```
yikes. so this means that idle threads are just going to spin wasting power? is there a bug filed on this?

abrodkin replied:
Let me check what's wrong here. I suggest to report it as a separate issue and continue discussion there so we track it for the next release etc.

ruuddw replied:
Interrupts should wake up the processor. Didn't check in detail, but is the z_arc_cpu_sleep_mode initialized correctly? It may set the threshold to mask the IPI interrupt.

abrodkin replied:
@ruuddw, @vonhust  `z_arc_cpu_sleep_mode` initialized with `_ARC_V2_WAKE_IRQ_LEVEL` = `_ARC_V2_DEF_IRQ_LEVEL`, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/arch/arc/include/v2/irq.h.

And `_ARC_V2_DEF_IRQ_LEVEL ` in absence of `CONFIG_ARC_SECURE_FIRMWARE` = `CONFIG_NUM_IRQ_PRIO_LEVELS - 1`, i. e. =1 for `nsim_hs_smp` & `hsdk` due to `NUM_IRQ_PRIO_LEVELS` = 2, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/soc/arc/snps_nsim/Kconfig.defconfig.hs_smp#L12 and https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/soc/arc/snps_arc_hsdk/Kconfig.defconfig#L19.

Inter-Core Interrupts are initialized with `ARCV2_ICI_IRQ_PRIORITY = 1`, see https://github.com/foss-for-synopsys-dwc-arc-processors/zephyr/blob/topic-fixes-smp/arch/arc/core/arc_smp.c#L24.

Now PRM says:
> Any interrupt with a higher priority than this interrupt threshold awakens the processor.

But later in an example I see:
>  SLEEP 0x014: Enter sleep and enable interrupts and sets interrupt priority level to 4. Any interrupt with a priority EQUAL TO or HIGHER THAN 4 awakens the processor.

Also I know that in the Linux kernel we rely on EQUAL TO part, i.e. we specify the IRQ level as when initializing all IRQs, see https://elixir.bootlin.com/linux/latest/source/arch/arc/kernel/process.c#L109 and that works. So I believe there should be something else.

ruuddw replied:
Thanks for checking - if the sleep works for Linux, it should work for Zephyr as well (i.e. not HW issue).

andyross replied:
FWIW: we merged SMP with a generic spin-for-idle implementation, and it was left that way for several months until the IPI framework landed.  And it was that change that broke SMP on ARC and is being addressed here.  Technically this isn't even a regression.

We should track it with a bug and fix it, but I don't see why this implementation can't merge.

vonhust replied:
For HW, e.g., HSDK,  sleep now works as expected  after i fixed the bug in systick driver (ipi interrupt is always preempted by sys tick int).  But in nsim_hs_smp, sleep still doesn't work as expected, it may be a nsim issue. Considering nsim is an instruction simulator,  there are few impact of using loop to replace sleep.

tyrantlizard77 reviewed (commented):
@andyross is the expert here, but meanwhile can you speak more about the issue where IPIs don't take the CPU out of low power wait instruction?

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -103,18 +72,60 @@ void z_arch_slave_start(int cpu_num)
 {
```
How `slave_start()` could be executed w/o SMP at all?

vonhust replied:
see tests/kernel/spinlock and tests/kernel/mp

abrodkin replied:
Well only in `tests/kernel/mp`! In `tests/kernel/spinlock` it's only referred in comments.

abrodkin commented on arch/arc/core/reset.S:
```diff
@@ -137,15 +137,14 @@ _slave_core_wait:
 #if defined(CONFIG_ARC_FIRQ_STACK)
 	jl z_arc_firq_stack_set
 #endif
+
```
Extra space!

vonhust replied:
will fix

abrodkin commented on arch/arc/core/arc_connect.c:
```diff
@@ -26,8 +26,15 @@ static struct k_spinlock arc_connect_spinlock;
+
 	LOCKED(&arc_connect_spinlock) {
-		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_GENERATE_IRQ, core);
+		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_READ_STATUS, core);
```
So now if for whatever reason other core is still busy processing another cross-core IRQ it won't get a new one and so effectively we'll forget to deliver some important piece of information to the remote core.

In that case we need to have a loop in which we'll be waiting until previous IRQ gets acknowledged.

vonhust replied:
I refer arc codes in linux to do this change.  It make sense to avoid uncessary operation of arc connect.  As the ipi from initiator to the receiver is already pending and consider current Zephyr's codes,  nothing will be missed.

vonhust replied:
I remove the change. it has a potential bug mentioned by @andyross

abrodkin reviewed (changes requested):
Even though series is pretty good I made a couple of comments in changes and also I think it should be split into 2 or 3 pull-requests.
1. Generic kernel changes
   * kernel: need to release spinlock before busy_wait
   * kernel: sched: do not abort a thread that's already dead
2. ARC-related changes
3. Test changes
   * tests: smp: fix the bug in test logic

Commit:
boards: nsim: add missed -firq for nsim_hs_smp

nsim_hs_smp has 2 reg banks, so it should have firq

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
arch: arc: clear ici interrupt during init

clear the ici interrupts during init

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

andyross commented on arch/arc/core/arc_smp.c:
```diff
@@ -44,15 +44,15 @@ u64_t z_arch_smp_switch_in_isr(void)
 	u32_t new_thread;
 	u32_t old_thread;
 
-	if (!_current_cpu->swap_ok) {
```
Agreed, not understanding this change.  The intent of "swap_ok" is that the current thread should be preemptible even if it is at a cooperative priority (the reason the flag has to exist in SMP is that the scheduler decision about the "next" thread to run cannot happen in the scheduler and needs to be done atomically at interrupt exit, so we need to save the decision).

This is a generic scheduler behavior, if it's going to be inspected from arch code it needs some kind of explanation as to why.

vonhust replied:
Condering ipi is used to send thread switch req through z_add_thread_to_ready_q.  It should not  rely swap->ok.  As @andyross  said, it's a generic scheduler behavior, no need to inspect from arch code.

The old codes are not correct.  So I remove it in this PR.   @andyross  could you give more background about swap_ok?

abrodkin commented on boards/arc/nsim/support/mdb_hs_smp.args:
```diff
@@ -23,6 +23,7 @@
 	-action_points=8
 	-Xstack_check
 	-interrupts=72
+	-firq
```
How does that related to SMP fixes?

vonhust replied:
The exception you found in nsim_hs_smp is caused by this.  mdb_hs_smp.arg is only for SMP.  so i think it can be part of SMP fixes

andyross commented on arch/arc/core/arc_smp.c:
```diff
@@ -44,15 +44,15 @@ u64_t z_arch_smp_switch_in_isr(void)
+/* should check whether new_thread to relace old_thread */
+#ifdef CONFIG_TIMESLICING
+		z_reset_time_slice();
+#endif
```
This code too looks like it's trying to duplicate logic that should be portable in the scheduler.  Actually lots of code around here does, predating this PR.  I think there may be a deeper problem here we may need to look at about why this had to be copied.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -76,6 +76,7 @@ void z_arch_slave_start(int cpu_num)
 	z_icache_setup();
 	z_irq_setup();
 
+	z_arc_connect_ici_clear();
```
I'd say it makes no sense. We assume that we start from clean platform, otherwise we'll end-up trying to init way too many things, some of which we don't have access to or don't now about.

In case of simulation we start from clean state anyway, in case of HW platform a rule of thumb is to reset your board before execution.

andyross replied:
FWIW, as someone not at all expert on the platform: warm reset conditions can routinely leave hardware in crazy states at startup.  This looks safe and harmless to me.

vonhust replied:
Yes, this part is to make code more safe.  As for hsdk and nsim,  in debug,  we just can make a warm reset.

abrodkin replied:
If we leave this I'd suggest to add a comment explaining what are we trying to achieve here and warn people about possible issues on warm reset.

abrodkin commented on arch/arc/core/arc_smp.c:
```diff
@@ -154,6 +155,7 @@ static int arc_smp_init(struct device *dev)
 
 	if (bcr.ipi) {
 	/* register ici interrupt, just need master core to register once */
+		z_arc_connect_ici_clear();
```
Ditto.

andyross commented on arch/arc/core/arc_smp.c:
```diff
@@ -120,7 +120,7 @@ void z_arch_sched_ipi(void)
 	u32_t i;
 
-	/* broadcast sched_ipi request to all cores
+	/* broadcast sched_ipi request to other cores
```
FWIW: the Zephyr sched_ipi API is agnostic as to whether or not the current CPU gets interrupted or not.  It's designed purposefully to be idempotent like this.  Either scheme is legal.

Personally, if this is a one-line change to a comment that reflects hardware behavior more precisely, I'm fine with it being added to a not-completely-related commit. :)

andyross commented on kernel/sched.c:
```diff
@@ -1065,17 +1068,19 @@ void z_sched_abort(struct k_thread *thread)
+			k_spin_unlock(&sched_spinlock, key);
+		} else {
+			k_spin_unlock(&sched_spinlock, key);
+			k_busy_wait(100);
```
There's another patch out there with exactly this change.  I agree with the logic, but have the same style comment here: can you leave the outer lock/unlock pair at the same level of indentation. and just put an unlock/lock around the k_busy_wait() call itself?

IMHO lock maintenance gets hard, and needs to be expressed in easily seen matched pairs wherever possible.

vonhust replied:
after #20046 is merged,  i will make a clean up of this PR.

andyross commented on arch/arc/core/arc_smp.c:
```diff
@@ -103,18 +72,60 @@ void z_arch_slave_start(int cpu_num)
+	if (new_thread != old_thread) {
+#ifdef CONFIG_TIMESLICING
+		z_reset_time_slice();
+#endif
```
This is another spot where you're doing something the scheduler should have already done for you.  Why is the time slice code needed here?

vonhust replied:
I think it may be required. each time when there is a thread switch , z_reset_time_slice should be called.   For SMP's case  z_reset_time_slice() will not be called in update_cache like single processor.

so consider a case of ipi interrupt.   One thread T2 is is made ready by T1 in core A through  z_add_thread_to_ready_q and will run in core B through ipi interrupt.  As there are no more higher priority ready  thread to run , then T1 in core A will continue to run, z_reset_time_slice will not be calledd because no swap in core A.   For core B, if  z_reset_time_slice is not called, i think there will be problem.

andyross commented on arch/arc/core/cpu_idle.S:
```diff
@@ -43,7 +43,18 @@ SECTION_FUNC(TEXT, z_arch_cpu_idle)
+	seti r1
+_z_arc_idle_loop:
+	b _z_arc_idle_loop
+#endif
```
Alternatively, you could elaborate the spin loop that already exists in kernel/idle.c so it's predicated on more than just CONFIG_SCHED_IPI_SUPPORTED and use that one instead.  Though spinning is just two instructions, so maybe it doesn't matter.

andyross commented on arch/arc/core/arc_connect.c:
```diff
@@ -26,8 +26,15 @@ static struct k_spinlock arc_connect_spinlock;
-		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_GENERATE_IRQ, core);
+		z_arc_connect_cmd(ARC_CONNECT_CMD_INTRPT_READ_STATUS, core);
+		ipi_was_pending = z_arc_connect_cmd_readback();
+		if (!ipi_was_pending) {
```
This looks super racy to me.  What if the IPI was handled between that line and the next?  The IPI won't be sent, and the kernel code on our core that requested it (e.g. trying to kill a thread) will spin until the next time the target core takes an interrupt or switches out of its thread, which can be a very long time.

Is there a hardware limitation that you can't flag colliding IPIs like this?  If there is, this might need more elaborate code to be 100% sure the IPI gets delivered.

vonhust replied:
make sense, i will think about it

andyross commented on kernel/sched.c:
```diff
@@ -1055,6 +1055,13 @@ void z_sched_abort(struct k_thread *thread)
+	 */
+	if (thread->base.thread_state & _THREAD_DEAD) {
+		return;
+	}
```
Is this really needed?  There is a loop checking for _THREAD_DEAD immediately after the IPI that follows this line, which should be guaranteed safe (the target thread obviously isn't running, it's dead).

Basically, this sounds like a pure optimization to me where we can avoid an IPI in cases where user code is trying to abort a thread that already exited.  Is that really a case worth optimizing?

If it's a bug fix instead, then we should document that.

vonhust replied:
This part will in a separete PR to trace. It's more like an optimization to avoid unnecessary ipi interrupt.  I am also thinking whether multiple abort ops on the thread will have side effects, especially in SMP.

andyross commented on tests/kernel/smp/src/main.c:
```diff
@@ -238,8 +238,14 @@ void test_coop_resched_threads(void)
+#else
 	spawn_threads(K_PRIO_COOP(10), THREADS_NUM, !EQUAL_PRIORITY,
 		      &thread_entry, THREAD_DELAY);
+#endif
```
Can you explain this one better?  I don't understand the bug being fixed.  And regardless: we don't want test code predicated on kernel features unless it's actively testing those features.

If we have test code where timing like this is causing failures, what we almost certainly have is a synchronization bug elsewhere in the test.  Can we fix that instead?

vonhust replied:
* one aspect is :
    For nsim_hs_smp,  i found in test_coop_resched_threads,  threads in other cores didn't run 
    even after k_busy_wait(),  it will make the following zassert_ture failed. 
    
    The key of test_coop_resched_threads is that other cores must run before k_busy_wait. But  this condition cannot be always met (i found both in nsim_hs_smp and hsdk).

So as you said,  we should have a better  synchronization here, but not rely on timing.

* another aspect is that
  After IPI is supported,  there is no need to do tiny delay  which is required by previous busy_wait in idle loop.

andyross reviewed (changes requested):
A few comments and confusions.   The only one I think is critically important is the apparent race with delivering an IPI, which needs to be fixed or explained for sure.

Commit:
drivers: timer: fix the bug for SMP

* fix the smp timer dirver bugs found in debug and test.
for smp case, GFRC is used as clock source, and local
internal timer is used to trigger time event.

* because 64-bits gfrc is used, so idle can be igored as no kernel
tick will be missed

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

Commit:
soc: arc: enable SYSTEM_CLOCK_SLOPPY for SMP

in SMP, arc uses gfrc as wall clock, so it's ok
to enable SYSTEM_CLOCK_SLOPPY to avoid unnecessary
interrupts

Signed-off-by: Wayne Ren <wei.ren@synopsys.com>

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -171,7 +171,6 @@ static void timer_int_handler(void *unused)
 
 	dticks = (curr_time - last_time) / CYC_PER_TICK;
 	last_time = curr_time;
-
```
Why doing that?

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -235,7 +234,7 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 	 */
 #ifdef CONFIG_SMP
-	if (IS_ENABLED(CONFIG_TICKLESS_IDLE) && idle && ticks == K_FOREVER) {
+	if (IS_ENABLED(CONFIG_TICKLESS_IDLE) && ticks == K_FOREVER) {
```
Could you please elaborate a bit more what was wrong here in the first place?
Otherwise how can we justify exactly the change you made?

vonhust replied:
For SMP, as we have gfrc, so no need to rely on idle.  What's more, if ticks ==K_FOREVER (-1) and idle == false,  then following codes can not work correctly

abrodkin replied:
1. How SMP & GFRC usage exclude a need of `idle`? It's not that I don't like that change but instead I'd just like to really understand what is done and why. Please pardon my noob questions but I'm sure that will simplify understanding rationale for your changes for other readers who may have less experience with Zephyr than you.

2. Whatever justification you have please put it in the commit message so via `git log` or `git blame` people may figure out what was changed and why.

vonhust replied:
already add extra comments in commit msg and code comments

abrodkin commented on drivers/timer/arcv2_timer0.c:
```diff
@@ -246,7 +245,8 @@ void z_clock_set_timeout(s32_t ticks, bool idle)
 
-	ticks = MIN(MAX_TICKS, MAX(ticks - 1, 0));
+	/* aligned to next target tick */
+	ticks = MIN(MAX_TICKS, ticks);
```
Here also it's completely unclear why do we do what we do and why new implementation is better. And even if that change helps us to get some test running better if we don't understand real root-cause of the issue we'll hit it later in more obscure case.

vonhust replied:
It's bug here, no need -1, because I copy old codes in no-smp case where there is operation of aligned to next tick (just like + 1) later.

abrodkin replied:
So is it a fatal bug and can we address it in a separate commit?

vonhust replied:
changes in timer driver are alread in a single commit  "drivers: timer: fix the bug for SMP".

as it's related to smp fix,  better in this PR.

abrodkin reviewed (changes requested):
Could you please explain clearly what was the issue we're trying to fix and how our changes fix whatever was broken?

vonhust commented on arch/arc/core/CMakeLists.txt:
```diff
@@ -27,4 +27,4 @@ add_subdirectory_ifdef(CONFIG_ARC_CORE_MPU mpu)
 zephyr_library_sources_ifdef(CONFIG_USERSPACE userspace.S)
 zephyr_library_sources_ifdef(CONFIG_ARC_CONNECT arc_connect.c)
-zephyr_library_sources_ifdef(CONFIG_SMP arc_smp.c)
+zephyr_library_sources_ifdef(CONFIG_ARC_CONNECT arc_smp.c)
```
The reason of change is that
* tests/kernel/spinlock and tests/kernel/mp  doesn't enable CONFIG_SMP but uses some multi-core functions like z_arch_slave_start() which depends on CONFIG_ARC_CONNECT.  
* I want arc_connect.c just has the arc connect related codes. so didn't put z_arch_slave_start etc. in arc_connect.c

* arc_smp.c  (maybe better name it arc_multicore) in fact has two parts
   * PART I:  codes are like baremetal multicore, e.g.  z_arch_start_cpu, z_arch_slave_start
   * PART II: codes are related to CONFIG_SMP

vonhust replied:
see tests/kernel/spinlock and tests/kernel/mp

abrodkin replied:
Ok agree we may consider to rename `arc_smp.c` to `arc_mp.c`. But then if there's no such thing as `CONFIG_MP` in Zephyr we may build-in `arc_*mp.c` unconditionally. Note `z_arch_start_cpu()` itself has nothing to do with ARConnect whatsoever.

abrodkin commented:
Ack for ARC-related changes.

vonhust commented:
@abrodkin @andyross @ruuddw  I made a clean up of this PR, just kept arc specific changes, removed changes in kernel and tests (will be in separate PR). So If no other comments,   let's get it merged.

ruuddw reviewed (approved):
I'm ok now, but please re-enable the tests/kernel/mp for ARC as this PR intends to have fixed that (in this PR, or a separate one).

abrodkin commented:
@vonhust I already approved this one so it's ready for merge I guess.
@nashif, @andyross?

andyross reviewed (approved):
OK, the race with the IPI delivery seems to be fixed, which was my only critical complaint.  Still don't love all the scheduler code that lives in the arch layer, but that's a cleanup thing.

tyrantlizard77 reviewed (changes requested):
> Still don't love all the scheduler code that lives in the arch layer, but that's a cleanup thing.

Which will never happen unless we file a bug on it.
Please, can we sort this out now? Either that or I want to see an issue filed.

andyross commented:
Fair enough, placeholder bug filed in #20671, we can add detail there as we discuss.

ruuddw commented:
Thanks Andy/Andrew, I agree further clean-up/simplification should be looked at, but that wasn't the purpose of this fix. @vonhust, please pick this up and work with Andy for next release.
